manager/controlapi/ca_rotation.go

```diff
@@ -0,0 +1,345 @@
+		var urls []string
+		for _, c := range cluster.Spec.CAConfig.ExternalCAs {
+			if c.Protocol == api.ExternalCA_CAProtocolCFSSL && bytes.Equal(c.CACert, cluster.RootCA.CACert) {
+				urls = append(urls, c.URL)
+			}
+		}
+		if len(urls) == 0 {
+			return nil, grpc.Errorf(codes.InvalidArgument,
+				"must provide an external CA for the current external root CA to generate a cross-signed certificate")
+		}
+		externalCA.UpdateURLs(urls...)
+		crossSignedCert, err = externalCA.CrossSignRootCA(ctx, newRootCA)
+	}
+
+	if err != nil {
+		log.G(ctx).WithError(err).Error("unable to generate a cross-signed certificate for root rotation")
+		return nil, grpc.Errorf(codes.Internal, "unable to generate a cross-signed certificate for root rotation")
+	}
+
+	copied := cluster.RootCA.Copy()
+	copied.RootRotation = &api.RootRotation{
+		CACert:            rootCert,
+		CAKey:             rootKey,
+		CrossSignedCACert: crossSignedCert,
+	}
+	copied.LastForcedRotation = version
+	return copied, nil
+}
+
+// TODO(cyli): in go 1.8 this will be easier as we can use *url.URL.Port() and *url.URL.Hostname()
+func getConnectionStringFromHost(hostport string) string {
+	colon := strings.IndexByte(hostport, ':')
+	if colon == -1 {
+		return fmt.Sprintf("%s:443", hostport) // no port, default to 443
+	}
+	// handle IPV6 addresses which look like http://[2001:db8:1f70::999:de8:7648:6e8]:8080/
+	if i := strings.IndexByte(hostport, ']'); i != -1 {
+		return strings.TrimPrefix(hostport[:i], "[") + hostport[colon:]
+	}
+	return hostport
+}
+
+// Checks that a CA URL is connectable using the credentials we have and that its server certificate is signed by the
+// root CA that we expect.
+func validateExternalCAURL(dialer *net.Dialer, tlsOpts *tls.Config, caURL string) error {
+	parsed, err := url.Parse(caURL)
+	if err != nil {
+		return err
+	}
+	if parsed.Scheme != "https" {
+		return errors.New("invalid HTTP scheme")
+	}
+	conn, err := tls.DialWithDialer(dialer, "tcp", getConnectionStringFromHost(parsed.Host), tlsOpts)
+	if conn != nil {
+		defer conn.Close()
+	}
+	return err
+}
+
+// Iterates over all the external CAs, and validates that there is at least 1 reachable, valid external CA for the
+// given CA certificate.  Returns true if there is, false otherwise.
+func hasAtLeastOneExternalCA(ctx context.Context, externalCAs []*api.ExternalCA, securityConfig *ca.SecurityConfig, wantedCert []byte) bool {
+	pool := x509.NewCertPool()
+	pool.AppendCertsFromPEM(wantedCert)
+	dialer := net.Dialer{Timeout: 5 * time.Second}
+	opts := tls.Config{
+		RootCAs:      pool,
+		Certificates: securityConfig.ClientTLSCreds.Config().Certificates,
+	}
+	for i, ca := range externalCAs {
+		if ca.Protocol == api.ExternalCA_CAProtocolCFSSL && bytes.Equal(wantedCert, ca.CACert) {
+			err := validateExternalCAURL(&dialer, &opts, ca.URL)
+			if err == nil {
+				return true
+			}
+			log.G(ctx).WithError(err).Debugf("external CA # %d is unreachable or invalid", i+1)
```

Comment:
I don't see any harm in warning every time the cluster is updated, as long as the unreachable CAs continue to be referenced in the spec. We're checking the CAs anyway, so if we discover that one is unreachable we might as well expose that information.